YatimaStdLib/List.lean

```diff
@@ -39,4 +39,83 @@ instance : Foldable List where
   foldr := List.foldr
   foldl := List.foldl
 
+-- the `B` suffix avoids name conflicts with mathlib
+def eraseDupB [BEq α] : List α → List α
+  | [] => []
+  | x::xs => 
+    let exs := eraseDupB xs
+    if exs.contains x then exs else x::exs
+
+def splitAtP [BEq α] (p : α → Bool) (l : List α) : List α × List α :=
+  match l.dropWhile p with 
+  | [] => unreachable!
+  | a::as => ⟨l.takeWhile p ++ [a], as⟩
+
+partial def mergeM [Monad μ] (cmp: α → α → μ Ordering) : List α → List α → μ (List α)
+  | as@(a::as'), bs@(b::bs') => do
+    if (← cmp a b) == Ordering.gt
+    then List.cons b <$> mergeM cmp as bs'
+    else List.cons a <$> mergeM cmp as' bs
+  | [], bs => return bs
+  | as, [] => return as
+
+def mergePairsM [Monad μ] (cmp: α → α → μ Ordering) : List (List α) → μ (List (List α))
```

Comment:
Superficially, a total function, generally, can't call a partial function and stay total.

Reply:
The keyword `partial` in Lean means that we're not going to prove termination for it. And then what Lean does is treat that function as a constant, whose definition can't be unfolded, meaning that we can't reason about its behavior.

This is confusing, but it's not the same thing as total/partial in mathematics